torch/optim/adadelta.py

```diff
@@ -66,13 +76,18 @@ def step(self, closure=None):
 
                 state['step'] += 1
 
-                if group['weight_decay'] != 0:
-                    grad = grad.add(group['weight_decay'], p.data)
+                if group['weight_decay'] != 0 and group['l2_reg']:
+                    grad.add_(group['weight_decay'], p.data)
 
                 square_avg.mul_(rho).addcmul_(1 - rho, grad, grad)
                 std = square_avg.add(eps).sqrt_()
                 delta = acc_delta.add(eps).sqrt_().div_(std).mul_(grad)
+
+                if group['weight_decay'] != 0 and not group['l2_reg']:
+                    p.data.add_(-group['weight_decay'], p.data)
```

Comment:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).